Improved main "content stripping" process

Shebo opened the pull request:
Added a new way for the app to identify and delete styles (mainly text highlights and text colors).
Photos, images and other visual data are no longer will be erased with highlights.

The solution is not perfect, and I tried my best with the usage of pypdf2 to identify only text highlights and not other imported styles from Word (line underlines, strikethroughs, tables, etc.)

You are welcome to tinker and change the logic values, but I think even in the current state it can bring real value for you and other students.

Shebo committed:
Replace removeImages with removeWordStyle

Add a new method that wouldn't remove all images and styles, but instead only remove text highlights and text colors

Shebo committed:
Refactor remover.py

Remove wildcard imports, standardize misc spaces and assignment statements

Shebo commented:
And another thing: I don't know how you are creating the .exe file, so I didn't create a new version of it from the source files and left it for you to do after merging.

omer-re commented:
Hi Boaz,
It sounds like you did a great job!
Do you maybe have any pictures of the before and after?

I am currently in the middle of my exams, and I will review your changes
and merge them in a few days.

sorry for the delay and thanks for contributing!


Omer

On Sat, Jul 18, 2020 at 4:48 PM Boaz Shebo Berger <notifications@github.com>
wrote:

> And another thing: I don't know how you are creating the .exe file, so I
> didn't create a new version of it from the source files and left it for you
> to do after merging.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fomer-re%2FPDF_Answers__Hider%2Fpull%2F3%23issuecomment-660485637&data=02%7C01%7Comer.re%40campus.technion.ac.il%7Ce82021ac25a14b0f995c08d82b214e09%7Cf1502c4cee2e411c9715c855f6753b84%7C1%7C0%7C637306769358840776&sdata=T261l%2BKa2bfP7GusBAI00w4iSA44LJEIh2q6FhwqI7w%3D&reserved=0>,
> or unsubscribe
> <https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAJKXIT4EI6SQWL3LJOGRJ73R4GR4BANCNFSM4O75WBEQ&data=02%7C01%7Comer.re%40campus.technion.ac.il%7Ce82021ac25a14b0f995c08d82b214e09%7Cf1502c4cee2e411c9715c855f6753b84%7C1%7C0%7C637306769358840776&sdata=atcQqBgBwVSYStD9HmNND8jZxDSS12%2FKzaI7UYVyz2w%3D&reserved=0>
> .
>
> *This email is from an external mail server, be judicious when opening
> attachments or links *
>

Shebo commented:
Hi Omer,
A friend gave 3-4 PDF files as an example, but I don't know what and if i can share the contents of those files.
If you wanna you can attach here a file of your choice and I can pass it through my version and take some screenshots/upload the "clean" version here.

I'm not a student myself so I don't have any access to those types of PDF files.

omer-re commented:
תרצה לעבור לעברית?
אני באמצע תקופת מבחנים אז אני ממש עוצר את עצמי לא להשאב לקוד הזה שוב, אבל
מאוד סקרן לראות מה עשית.

צרפתי קובץ לדוגמה

On Mon, Jul 20, 2020 at 8:20 PM Boaz Shebo Berger <notifications@github.com>
wrote:

> Hi Omer,
> A friend gave 3-4 PDF files as an example, but I don't know what and if i
> can share the contents of those files.
> If you wanna you can attach here a file of your choice and I can pass it
> through my version and take some screenshots/upload the "clean" version
> here.
>
> I'm not a student myself so I don't have any access to those types of PDF
> files.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fomer-re%2FPDF_Answers__Hider%2Fpull%2F3%23issuecomment-661210513&data=02%7C01%7Comer.re%40campus.technion.ac.il%7C42ca446501344c52ee9808d82cd132cf%7Cf1502c4cee2e411c9715c855f6753b84%7C1%7C0%7C637308624298336360&sdata=oJ2bHruI0YtWcIAaNWRzOOxDp4%2BN6PEtvyCxtbYlYVg%3D&reserved=0>,
> or unsubscribe
> <https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAJKXIT3TMPBH5B2CZ6FRYM3R4R4FXANCNFSM4O75WBEQ&data=02%7C01%7Comer.re%40campus.technion.ac.il%7C42ca446501344c52ee9808d82cd132cf%7Cf1502c4cee2e411c9715c855f6753b84%7C1%7C0%7C637308624298346356&sdata=z7h6F1eBN7vmj3kj6DAgyZBNCzYM%2BzqWxJwOJSTFYj4%3D&reserved=0>
> .
>
> *This email is from an external mail server, be judicious when opening
> attachments or links *
>

Shebo commented:
You can feel free to write in Hebrew, I'm more accustomed to English :)
I don't see any attached file, you can just upload it to some storage service and share the link here.

omer-re commented:
trying again, lmk if you got it
https://we.tl/t-hWdimK7Yu8

On Mon, Jul 20, 2020 at 10:54 PM Boaz Shebo Berger <notifications@github.com>
wrote:

> You can feel free to write in Hebrew, I'm more accustomed to English :)
> I don't see any attached file, you can just upload it to some storage
> service and share the link here.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fomer-re%2FPDF_Answers__Hider%2Fpull%2F3%23issuecomment-661300139&data=02%7C01%7Comer.re%40campus.technion.ac.il%7Cfad7c71873ce47b490a708d82ce6c54c%7Cf1502c4cee2e411c9715c855f6753b84%7C1%7C0%7C637308716956880552&sdata=urjF5LLRz4t6n%2F0sb9dqXNQpMeAWQdFH7MMDQH4h2Go%3D&reserved=0>,
> or unsubscribe
> <https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAJKXITZGVAQCX5QOUZHIBKTR4SOIPANCNFSM4O75WBEQ&data=02%7C01%7Comer.re%40campus.technion.ac.il%7Cfad7c71873ce47b490a708d82ce6c54c%7Cf1502c4cee2e411c9715c855f6753b84%7C1%7C0%7C637308716956880552&sdata=9Qpe1qGpJIG%2BaSUnlAhdWP5ddk09xuOODm%2Byfo8WLv0%3D&reserved=0>
> .
>
> *This email is from an external mail server, be judicious when opening
> attachments or links *
>

Shebo committed:
Change color logics in removeWordStyle

The previous flow resulted in some missed highlight blocks and some highlights turning full black blocks.
The new flow removed the graphic sequence check completely since I found many colors and highlights happens outside the sequence.
In addition, I added a check to determine if color is used by a text or rectangle object. All texts are colored black, and all rectangles are colored white. This will guarantee that even rectangle blocks that are not deleted will be invisible(white).

Shebo committed:
Add line content support and remove seq_graphics

Add support to line objects. Some color operations are in relation to stroked lines, not just rectangles and texts.
Remove the use of graphics sequence. The implementation is buggy and the tested doc showed that graphic elements can be outside the said sequence.

Shebo commented:
OK, so at first the PDF that you sent didn't work as expected.
The problem is that every doc that I tested was structured differently from the others. That caused me to rewrite the logic and flow of the function every time from the start.

This happens mainly because the element (re, rectangle) who is used to represent highlights is used to represent other graphical elements (like underlines, borders, tables, etc.).

I've rewritten the function again, identifying some problems and improving upon the existing logic. The newer version works with the example doc that you shared, but because every doc is built slightly differently, I think it's best to test as many as possible.

The current logic, like the ones before, is not full-proof, but it's getting better. The more testing we'll do - the more it will improve.

### **The clean document**
[example_src-clean.pdf](https://github.com/omer-re/PDF_Answers__Hider/files/4960238/example_src-clean.pdf)

### **Screenshots**
Original Document
:-------------------------:
![](https://user-images.githubusercontent.com/299232/88181808-0ad22300-cc38-11ea-888e-0414d440dc0c.JPG)

Original Version |  New version A | New Version B
:-------------------------:|:-------------------------:|:-------------------------:
![](https://user-images.githubusercontent.com/299232/88181809-0c035000-cc38-11ea-8b57-ec9c4eefa020.JPG)  |  ![](https://user-images.githubusercontent.com/299232/88181811-0c9be680-cc38-11ea-8d41-2018cbdd7563.JPG)  |  ![](https://user-images.githubusercontent.com/299232/88181815-0c9be680-cc38-11ea-9ebe-75de89941679.JPG)

omer-re commented:
this is all very impressive, and I am super flattered you have dedicated
your time for improving my project.
I was very surprised that nothing similar is already exists and I am glad
it got a few hundred people's attention.

On Wed, Jul 22, 2020 at 4:40 PM Boaz Shebo Berger <notifications@github.com>
wrote:

> OK, so at first the PDF that you sent didn't work as expected.
> The problem is that every doc that I tested was structured differently
> from the others. That caused me to rewrite the logic and flow of the
> function every time from the start.
>
> This happens mainly because the element (re, rectangle) who is used to
> represent highlights is used to represent other graphical elements (like
> underlines, borders, tables, etc.).
>
> I've rewritten the function again, identifying some problems and improving
> upon the existing logic. The newer version works with the example doc that
> you shared, but because every doc is built slightly differently, I think
> it's best to test as many as possible.
>
> The current logic, like the ones before, is not full-proof, but it's
> getting better. The more testing we'll do - the more it will improve.
> *The clean document*
>
> example_src-clean.pdf
> <https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fomer-re%2FPDF_Answers__Hider%2Ffiles%2F4960238%2Fexample_src-clean.pdf&data=02%7C01%7Comer.re%40campus.technion.ac.il%7C3e91bd1567964e3f307d08d82e44ce84%7Cf1502c4cee2e411c9715c855f6753b84%7C1%7C0%7C637310220915714373&sdata=IietLPerw5G0BoLZ1fNWDzgqdAN9fwH%2BvFzVFxtheUU%3D&reserved=0>
> *Screenshots*
> Original Document
>
> <https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F299232%2F88181808-0ad22300-cc38-11ea-888e-0414d440dc0c.JPG&data=02%7C01%7Comer.re%40campus.technion.ac.il%7C3e91bd1567964e3f307d08d82e44ce84%7Cf1502c4cee2e411c9715c855f6753b84%7C1%7C0%7C637310220915724369&sdata=e352QdHLdzpoFLK6YyfsRG6pEtHqWy7Ke%2Fy18j3ipU8%3D&reserved=0>
> Original Version New version A New Version B
>
> <https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F299232%2F88181809-0c035000-cc38-11ea-8b57-ec9c4eefa020.JPG&data=02%7C01%7Comer.re%40campus.technion.ac.il%7C3e91bd1567964e3f307d08d82e44ce84%7Cf1502c4cee2e411c9715c855f6753b84%7C1%7C0%7C637310220915734358&sdata=mYBKJ5yxuyiPUiRxpGgJND9lVX5yTftvusZdCdbpeNo%3D&reserved=0>
> <https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F299232%2F88181811-0c9be680-cc38-11ea-8d41-2018cbdd7563.JPG&data=02%7C01%7Comer.re%40campus.technion.ac.il%7C3e91bd1567964e3f307d08d82e44ce84%7Cf1502c4cee2e411c9715c855f6753b84%7C1%7C0%7C637310220915734358&sdata=qb4L7JjqdfFzLstXUqIkOYZ1SOKWiDqquNtz%2FrNSMQs%3D&reserved=0>
> <https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F299232%2F88181815-0c9be680-cc38-11ea-9ebe-75de89941679.JPG&data=02%7C01%7Comer.re%40campus.technion.ac.il%7C3e91bd1567964e3f307d08d82e44ce84%7Cf1502c4cee2e411c9715c855f6753b84%7C1%7C0%7C637310220915744357&sdata=JVj%2FaTdNU5hRFie8iYt9Lzbjh2ee271N3II%2F1UtqZq0%3D&reserved=0>
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fomer-re%2FPDF_Answers__Hider%2Fpull%2F3%23issuecomment-662459494&data=02%7C01%7Comer.re%40campus.technion.ac.il%7C3e91bd1567964e3f307d08d82e44ce84%7Cf1502c4cee2e411c9715c855f6753b84%7C1%7C0%7C637310220915754347&sdata=4e2fA9OeeOIJnd%2BEtFBFVrRe7aoNvzbG6VKDY1MAP7w%3D&reserved=0>,
> or unsubscribe
> <https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAJKXIT4772OAEXL2BXWFRLTR43T47ANCNFSM4O75WBEQ&data=02%7C01%7Comer.re%40campus.technion.ac.il%7C3e91bd1567964e3f307d08d82e44ce84%7Cf1502c4cee2e411c9715c855f6753b84%7C1%7C0%7C637310220915754347&sdata=M2MnaqLkM4%2FVNZZfyv5U%2BUf%2FhvwHdTyWGOy85W2odE4%3D&reserved=0>
> .
>
> *This email is from an external mail server, be judicious when opening
> attachments or links *
>

Shebo commented:
> this is all very impressive, and I am super flattered you have dedicated your time for improving my project. I was very surprised that nothing similar is already exists and I am glad it got a few hundred people's attention.

Thank you for the kind words, I'm just happy to help. I love it when people use programming to make everyday life a little bit easier :)
I'm sure that many students will find this app very useful to them.

If any bugs or unintended behavior will come up in the future don't hesitate to ping me, I'd be happy to help.
